contracts/src/rng/BlockhashRNG.sol

```diff
@@ -1,64 +1,50 @@
+ *  This contract must be used when returning 0 is a worse failure mode than returning another blockhash.
+ *  Allows saving the random number for use in the future. It allows the contract to still access the blockhash even after 256 blocks.
  */
 contract BlockHashRNG is RNG {
-    mapping(uint256 => uint256) public randomNumber; // randomNumber[block] is the random number for this block, 0 otherwise.
-    mapping(uint256 => uint256) public reward; // reward[block] is the amount to be paid to the party w.
-
-    /** @dev Contribute to the reward of a random number.
-     *  @param _block Block the random number is linked to.
-     */
-    function contribute(uint256 _block) public payable override {
-        reward[_block] += msg.value;
-    }
+    mapping(uint256 => uint256) public randomNumbers; // randomNumbers[block] is the random number for this block, 0 otherwise.
 
-    /** @dev Return the random number. If it has not been saved and is still computable compute it.
+    /**
+     *  @dev Request a random number.
      *  @param _block Block the random number is linked to.
-     *  @return RN Random Number. If the number is not ready or has not been requested 0 instead.
      */
-    function getRN(uint256 _block) public override returns (uint256 RN) {
-        RN = randomNumber[_block];
-        if (RN == 0) {
-            saveRN(_block);
-            return randomNumber[_block];
-        } else {
-            return RN;
-        }
+    function requestRandomness(uint256 _block) external override {
+        // nop
     }
 
-    /** @dev Save the random number for this blockhash and give the reward to the caller.
+    /**
+     *  @dev Return the random number. If it has not been saved and is still computable compute it.
      *  @param _block Block the random number is linked to.
+     *  @return randomNumber The random number or 0 if it is not ready or has not been requested.
      */
-    function saveRN(uint256 _block) public virtual {
-        if (blockhash(_block) != 0x0) {
-            randomNumber[_block] = uint256(blockhash(_block));
+    function receiveRandomness(uint256 _block) external override returns (uint256 randomNumber) {
+        randomNumber = randomNumbers[_block];
+        if (randomNumber != 0) {
+            return randomNumber;
         }
 
-        if (randomNumber[_block] != 0) {
-            // If the number is set.
-            uint256 rewardToSend = reward[_block];
-            reward[_block] = 0;
-            payable(msg.sender).send(rewardToSend); // Note that the use of send is on purpose as we don't want to block in case msg.sender has a fallback issue.
+        if (_block < block.number && randomNumbers[_block] == 0) {
+            // The random number is not already set and can be.
+            if (blockhash(_block) != 0x0) {
+                // Normal case.
+                randomNumbers[_block] = uint256(blockhash(_block));
```

Comment:
not a big deal, but would be cheaper to do 
`randomNumber = uint256(blockhash(_block));`
`randomNumbers[_block] = randomNumber;`

and remove line 48. Similarly for the else block.

`randomNumber = uint256(blockhash(block.number - 1));`
`randomNumbers[_block] = randomNumber;`